### PR TITLE
Switch BIP340 to even R tiebreaker

### PR DESCRIPTION
**Pull request by sipa**
Also change the tags from "BIP340/*" to "BIP0340/*" to avoid collisions with earlier drafts.

### CODE REVIEW

File: bip-0340.mediawiki [hunk 1]
```diff
@@ -217,20 +206,17 @@ The algorithm ''BatchVerify(pk<sub>1..u</sub>, m<sub>1..u</sub>, sig<sub>1..u</s
...
-
-'''Squareness testing''' The function ''is_square(x)'' is defined as above, but can be computed more efficiently using an [https://en.wikipedia.org/wiki/Jacobi_symbol#Calculating_the_Jacobi_symbol extended GCD algorithm].
+Many techniques are known for optimizing elliptic curve implementations. Several of them apply here, but are out of scope for this document. One is listed below however, as it is relevant to the design decisions:
 
 '''Jacobian coordinates''' Elliptic Curve operations can be implemented more efficiently by using [https://en.wikibooks.org/wiki/Cryptography/Prime_Curve/Jacobian_Coordinates Jacobian coordinates]. Elliptic Curve operations implemented this way avoid many intermediate modular inverses (which are computationally expensive), and the scheme proposed in this document is in fact designed to not need any inversions at all for verification. When operating on a point ''P'' with Jacobian coordinates ''(x,y,z)'' which is not the point at infinity and for which ''x(P)'' is defined as ''x / z<sup>2</sup>'' and ''y(P)'' is defined as ''y / z<sup>3</sup>'':
```
Comment (real-or-random) [inline bip-0340.mediawiki:219]: Here we say that the scheme is designed to not need any inversions at all. This is not really true anymore.

File: bip-0340.mediawiki [hunk 2]
```diff
@@ -217,20 +206,17 @@ The algorithm ''BatchVerify(pk<sub>1..u</sub>, m<sub>1..u</sub>, sig<sub>1..u</s
...
 * Return success iff no failure occurred before reaching this point.
 
 If all individual signatures are valid (i.e., ''Verify'' would return success for them), ''BatchVerify'' will always return success. If at least one signature is invalid, ''BatchVerify'' will return success with at most a negligible probability.
 
 === Optimizations ===
 
-Many techniques are known for optimizing elliptic curve implementations. Several of them apply here, but are out of scope for this document. Two are listed below however, as they are relevant to the design decisions:
-
-'''Squareness testing''' The function ''is_square(x)'' is defined as above, but can be computed more efficiently using an [https://en.wikipedia.org/wiki/Jacobi_symbol#Calculating_the_Jacobi_symbol extended GCD algorithm].
+Many techniques are known for optimizing elliptic curve implementations. Several of them apply here, but are out of scope for this document. One is listed below however, as it is relevant to the design decisions:
```
Comment (real-or-random) [inline bip-0340.mediawiki:217]: Are Jacobian coordinates still relevant to the design decisions? If no, do we need this entire section at all?

File: bip-0340.mediawiki [hunk 3]
```diff
@@ -78,14 +78,9 @@ Using the first option would be slightly more efficient for verification (around
...
+The second option offers the greatest compatibility with existing key generation systems, where the standard 33-byte compressed public key format consists of a byte indicating the oddness of the Y coordinate, plus the full X coordinate. To avoid gratuitous incompatibilities, we pick that option for ''P'', and thus our X-only public keys become equivalent to a compressed public key that is the x-only key prefixed by the byte 0x02. For consistency, the same is done for ''R''<ref>An earlier version of this draft used the third option instead, based on a belief that this would in general trade signing efficiency for verification efficiency. Determining if a Y coordinate is a square can be done by computing the Legendre symbol directly in Jacobian coordinates (a common optimization), while determining oddness needs a conversion to affine coordinates first, which needs a modular inversion. Recent [https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-August/018081.html work] shows that a modular inverse can be computed faster than Legendre symbol, and thus conversion to affine coordinates can be faster than determining oddness.</ref>.
```
[the remark this answers is not in the document]
Comment (sipa) [reply bip-0340.mediawiki]: Changed everything to X-only.

File: bip-0340.mediawiki (hunk 1 above)
Comment (sipa) [reply bip-0340.mediawiki]: Fixed by dropping the section entirely.

File: bip-0340.mediawiki (hunk 2 above)
Comment (sipa) [reply bip-0340.mediawiki]: I guess it isn't. I've dropped it.

File: bip-0340.mediawiki [hunk 4]
```diff
@@ -75,27 +75,22 @@ We note that key prefixing is not strictly necessary for transaction signatures
...
+The second option offers the greatest compatibility with existing key generation systems, where the standard 33-byte compressed public key format consists of a byte indicating the oddness of the Y coordinate, plus the full X coordinate. To avoid gratuitous incompatibilities, we pick that option for ''P'', and thus our X-only public keys become equivalent to a compressed public key that is the X-only key prefixed by the byte 0x02. For consistency, the same is done for ''R''<ref>An earlier version of this draft used the third option instead, based on a belief that this would in general trade signing efficiency for verification efficiency. Determining if a Y coordinate is a square can be done by computing the Legendre symbol directly in Jacobian coordinates (a common optimization), while determining oddness needs a conversion to affine coordinates first, which needs a modular inversion. Recent [https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-August/018081.html work] shows that a modular inverse can be computed faster than Legendre symbol, and thus conversion to affine coordinates can be faster than determining oddness.</ref>.
```
Comment (jonasnick) [inline bip-0340.mediawiki:85]: nit: I wouldn't mention the recent work because it seems irrelevant. What matters is that Legendre symbol and affine conversion have *similar* computational cost with modern algorithms. Currently it sounds a bit like the recent work was essential for revising the draft. I'd still link to the mailing list post.

File: bip-0340.mediawiki [hunk 5]
```diff
@@ -75,27 +75,22 @@ We note that key prefixing is not strictly necessary for transaction signatures
 
 Using the first option would be slightly more efficient for verification (around 10%), but we prioritize compactness, and therefore choose option 3.
 
+Despite halving the size of the set of valid public keys, implicit Y coordinates are not a reduction in security. Informally, if a fast algorithm existed to compute the discrete logarithm of an X-only public key, then it could also be used to compute the discrete logarithm of a full public key: apply it to the X coordinate, and then optionally negate the result. This shows that breaking an X-only public key can be at most a small constant term faster than breaking a full one.<ref>This can be formalized by a simple reduction that reduces an attack on Schnorr signatures with implicit Y coordinates to an attack to Schnorr signatures with explicit Y coordinates. The reduction works by reencoding public keys and negating the result of the hash function, which is modeled as random oracle, whenever the challenge public key has an explicit Y coordinate that is odd. A proof sketch can be found [https://medium.com/blockstream/reducing-bitcoin-transaction-sizes-with-x-only-pubkeys-f86476af05d7 here].</ref>.
```
Comment (jonasnick) [inline bip-0340.mediawiki:78]: This doesn't seem to be the right place still, because the `'''implicit Y coordinates'''` section starts just below this paragraph. Perhaps move it right below the first sentence of the `implicit Y coordinates` section.

File: bip-0340.mediawiki [hunk 6]
```diff
@@ -213,26 +202,16 @@ Input:
 The algorithm ''BatchVerify(pk<sub>1..u</sub>, m<sub>1..u</sub>, sig<sub>1..u</sub>)'' is defined as:
 * Generate ''u-1'' random integers ''a<sub>2...u</sub>'' in the range ''1...n-1''. They are generated deterministically using a [https://en.wikipedia.org/wiki/Cryptographically_secure_pseudorandom_number_generator CSPRNG] seeded by a cryptographic hash of all inputs of the algorithm, i.e. ''seed = seed_hash(pk<sub>1</sub>..pk<sub>u</sub> || m<sub>1</sub>..m<sub>u</sub> || sig<sub>1</sub>..sig<sub>u</sub> )''. A safe choice is to instantiate ''seed_hash'' with SHA256 and use [https://tools.ietf.org/html/rfc8439 ChaCha20] with key ''seed'' as a CSPRNG to generate 256-bit integers, skipping integers not in the range ''1...n-1''.
 * For ''i = 1 .. u'':
-** Let ''P<sub>i</sub> = lift_x_even_y(int(pk<sub>i</sub>))''; fail if it fails.
+** Let ''P<sub>i</sub> = lift_x(int(pk<sub>i</sub>))''; fail if it fails.
```
Comment (jonasnick) [inline bip-0340.mediawiki:205]: we need to rename that in bip-341 as well

File: bip-0340/test-vectors.py [hunk 7]
```diff
@@ -1,6 +1,16 @@
...
 from reference import *
 
+def is_square(x):
+    return int(pow(x, (p - 1) // 2, p)) == 1
+
+def has_square_y(P):
+    """Determine if P has a square Y coordinate. Used in an earlier draft of BIP340."""
+    infinity = is_infinity(P)
+    if infinity: return False
+    assert P is not None
```
Comment (jonasnick) [inline bip-0340/test-vectors.py:11]: without the mypy types this looks very weird. Perhaps just
```
"""<comment>"""
assert not is_infinity(P)
return is_square(P[1])

**Review by jonasnick — COMMENTED** [review]
I ran the reference code and reproduced the test vectors. I'll cross check the test vectors with the libsecp implementation today.

File: bip-0340.mediawiki (hunk 6 above)
Comment (sipa) [reply bip-0340.mediawiki:205]: Done.

File: bip-0340/test-vectors.py (hunk 7 above)
Comment (sipa) [reply bip-0340/test-vectors.py]: Done.

File: bip-0340.mediawiki (hunk 5 above)
Comment (sipa) [reply bip-0340.mediawiki]: You're right that it shouldn't talk about implicit Y coordinates, but on the other hand, the security argument doesn't really rely on that (and would equally apply to a hypothetical scheme where the Y coordinate is allowed to either).

I've dropped the commit for now, and we can discuss phrasing separately. It's unrelated to this change anyway.

File: bip-0340.mediawiki (hunk 4 above)
Comment (sipa) [reply bip-0340.mediawiki]: Indeed, that was giving the wrong impression. Updated it.

**Review by jonasnick — COMMENTED** [review]
Can confirm that if only  the tag is replaced with BIP0340, the libsecp schnorrsig implementation fails signing & verification at the first test vector. The test vectors pass after changing R's tiebreaker to evenness.

File: bip-0340.mediawiki [hunk 8]
```diff
@@ -78,14 +78,9 @@ Using the first option would be slightly more efficient for verification (around
 '''Implicit Y coordinates''' In order to support efficient verification and batch verification, the Y coordinate of ''P'' and of ''R'' cannot be ambiguous (every valid X coordinate has two possible Y coordinates). We have a choice between several options for symmetry breaking:
 # Implicitly choosing the Y coordinate that is in the lower half.
 # Implicitly choosing the Y coordinate that is even<ref>Since ''p'' is odd, negation modulo ''p'' will map even numbers to odd numbers and the other way around. This means that for a valid X coordinate, one of the corresponding Y coordinates will be even, and the other will be odd.</ref>.
-# Implicitly choosing the Y coordinate that is a quadratic residue (has a square root modulo the field size, or "is a square" for short)<ref>A product of two numbers is a square when either both or none of the factors are squares. As ''-1'' is not a square modulo secp256k1's field size ''p'', and the two Y coordinates corresponding to a given X coordinate are each other's negation, this means exactly one of the two must be a square.</ref>.
+# Implicitly choosing the Y coordinate that is a quadratic residue (i.e. has a square root).
```
Comment (real-or-random) [inline bip-0340.mediawiki:81]: Maybe bring back the "modulo the field order" after "square root", in particular because the other two options view the y coordinate as an integer and not a field element (how would you defined lower half or odd otherwise?)

File: bip-0340.mediawiki [hunk 9]
```diff
@@ -78,14 +78,9 @@ Using the first option would be slightly more efficient for verification (around
...
+The second option offers the greatest compatibility with existing key generation systems, where the standard 33-byte compressed public key format consists of a byte indicating the oddness of the Y coordinate, plus the full X coordinate. To avoid gratuitous incompatibilities, we pick that option for ''P'', and thus our X-only public keys become equivalent to a compressed public key that is the X-only key prefixed by the byte 0x02. For consistency, the same is done for ''R''<ref>An earlier version of this draft used the third option instead, based on a belief that this would in general trade signing efficiency for verification efficiency. Determining if a Y coordinate is a square can be done by computing the Legendre symbol directly in Jacobian coordinates (a common optimization), while determining oddness needs a conversion to affine coordinates first, which needs a modular inversion. As modular inverses and Legendre symbols have similar [https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-August/018081.html performance] in practice, this trade-off is not worth it.</ref>.
```
Comment (real-or-random) [inline bip-0340.mediawiki:83]: I agree, this is better. :)

s/is a square/is a quadratic residue
(We don't define "square" anymore)

**Review by real-or-random — COMMENTED** [review]
ACK on the text changes except nits, I still need to look at the code and vectors

File: bip-0340.mediawiki (hunk 9 above)
Comment (sipa) [reply bip-0340.mediawiki:83]: Done.

File: bip-0340.mediawiki (hunk 8 above)
Comment (sipa) [reply bip-0340.mediawiki]: Good point, done.

**Review by jonasnick — APPROVED** [review]
ACK

**Comment by ajtowns** [comment]
Bit late, but ACK
